packages/code-editor/src/CodeEditor/stories/Xml.tsx

```diff
@@ -0,0 +1,322 @@
+    </div>
+    <div className={classes.actionContainer}>
+      <HvButton variant="primaryGhost" onClick={onClickSearch}>
+        Search
+      </HvButton>
+      <HvButton variant="primaryGhost" onClick={onClickTree}>
+        XML Tree
+      </HvButton>
+    </div>
+  </div>
+);
+
+type Tree = Record<string, object | object[] | string | undefined>;
+
+type Attributes = Record<string, Record<string, string>>;
+
+interface SimpleTreeItemProps extends HvTreeItemProps {
+  attributes: Attributes;
+}
+
+const SimpleTreeItem = forwardRef<HTMLLIElement, SimpleTreeItemProps>(
+  (props, ref) => {
+    const { children, nodeId, label, attributes, ...others } = props;
+    return (
+      <HvTreeItem
+        ref={ref}
+        nodeId={nodeId}
+        label={
+          children ? (
+            <div className={classes.label}>
+              <HvTypography variant="label">{label}</HvTypography>
+              {attributes[nodeId] && (
+                <HvTooltip
+                  title={Object.entries(attributes[nodeId])
+                    .map(([key, value]) => `${key}: ${value}`)
+                    .join("; ")}
+                >
+                  <Info />
+                </HvTooltip>
+              )}
+            </div>
+          ) : (
+            label
+          )
+        }
+        {...others}
+      >
+        {children}
+      </HvTreeItem>
+    );
+  },
+);
+
+const renderItem = (
+  key: string,
+  value: object | object[] | string | undefined,
+  attributes: Attributes,
+) => {
+  const label = key.split("_")[0];
+
+  if (Array.isArray(value)) {
+    return (
+      <SimpleTreeItem
+        key={key}
+        nodeId={key}
+        label={label}
+        attributes={attributes}
+      >
+        {value.map((childValue) => {
+          const childKey = Object.keys(childValue)[0];
+          return renderItem(childKey, childValue[childKey], attributes);
+        })}
+      </SimpleTreeItem>
+    );
+  }
+
+  if (typeof value === "object") {
+    return (
+      <SimpleTreeItem
+        key={key}
+        nodeId={key}
+        label={label}
+        attributes={attributes}
+      >
+        {Object.entries(value).map(([childKey, childValue]) =>
+          renderItem(childKey, childValue, attributes),
+        )}
+      </SimpleTreeItem>
+    );
+  }
+
+  return (
+    <SimpleTreeItem
+      key={key}
+      nodeId={key}
+      label={label}
+      attributes={attributes}
+    >
+      {value}
+    </SimpleTreeItem>
+  );
+};
+
+export const XmlStory = () => {
+  const [opened, setOpened] = useState(false);
+  const [editorValue, setEditorValue] = useState(defaultValue);
+  const [xmlTree, setXmlTree] = useState<{
+    tree: Tree;
+    attributes: Attributes;
+  }>();
+  const [defaultExpandedKeys, setDefaultExpandedKeys] = useState<string[]>([]);
+
+  const editorRef = useRef<any>(null);
+
+  const buildTree = (xml: string) => {
+    const parser = new DOMParser();
+    const xmlDoc = parser.parseFromString(xml, "application/xml");
```

Comment:
This looks like a pure function. Being that the case, we should hoist this outside the component.

The TreeView looks nice, but it adds lots of complexity to the sample 😬
... but by hoisting this I think out of the component, I think it'll be easier to follow 👍